content/advent-2018/montecarlo.md

```diff
@@ -0,0 +1,829 @@
+
+The `wplot` type is just a shim to hold the resulting `base64` encoded string of the PNG plot, created with `renderImg`:
+[embedmd]:# (./montecarlo/mc-pi-plot.go go /type wplot/ /^}/)
+```go
+type wplot struct {
+	Plot string `json:"plot"`
+}
+```
+[embedmd]:# (./montecarlo/mc-pi-plot.go go /func renderImg/ /^}/)
+```go
+func renderImg(p *hplot.Plot) string {
+	size := 20 * vg.Centimeter
+	canvas := vgimg.PngCanvas{vgimg.New(size, size)}
+	p.Draw(draw.New(canvas))
+	out := new(bytes.Buffer)
+	_, err := canvas.WriteTo(out)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return base64.StdEncoding.EncodeToString(out.Bytes())
+}
+```
+
+At the other end of the `srv.plots` channel is the `dataHandler` method that pulls plots out to send them to the web client, over a WebSocket:
+[embedmd]:# (./montecarlo/mc-pi-plot.go go /func .* dataHandler/ /^}/)
+```go
+func (srv *server) dataHandler(ws *websocket.Conn) {
+	for data := range srv.plots {
+		err := websocket.JSON.Send(ws, data)
+		if err != nil {
+			log.Printf("error sending data: %v\n", err)
+		}
+	}
+}
+```
+
+and, *finally*, the `/` end point and its home page:
+[embedmd]:# (./montecarlo/mc-pi-plot.go go /func plotHandle/ /^}/)
+```go
+func plotHandle(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, page)
+}
+```
+[embedmd]:# (./montecarlo/mc-pi-plot.go go /const page/ /^\`/)
+```go
+const page = `
+<html>
+	<head>
+		<title>Monte Carlo</title>
+		<script type="text/javascript">
+		var sock = null;
+		var plot = "";
+
+		function update() {
+			var p = document.getElementById("plot");
+			p.src = "data:image/png;base64,"+plot;
+		};
+
+		window.onload = function() {
+			sock = new WebSocket("ws://"+location.host+"/data");
+
+			sock.onmessage = function(event) {
+				var data = JSON.parse(event.data);
+				plot = data.plot;
+				update();
+			};
+		};
+
+		</script>
+	</head>
+
+	<body>
+		<div id="content">
+			<p style="text-align:center;">
+				<img id="plot" src="" alt="Not Available"></img>
+			</p>
+		</div>
+	</body>
+</html>
+`
+```
+
+Running the Go program and navigating to `localhost:8080`, you should see:
+
+```
+$> go run ./mc-pi-plot.go
+monte-carlo: listening on :8080...
+```
+
+![plot-gauss](/postimages/advent-2018/montecarlo/mc-pi.png)
+
+## Conclusions
+
+I hope this quick foray into a technique that is at the heart of many physics simulations was fun.
+The Monte Carlo technique isn't always the fastest technique to perform simulations (this obviously depends on the "shape" of the function we want to model) but it is deceptively simple, and can be visually "fun" -- for some definition of _fun_.
+
+Note that in this little example, we had fun creating a PNG image, encoding it to JSON+`base64` and sending it over a WebSocket.
+This was just to exercize a bunch of packages with Go.
+[Gonum/plot](https://godoc.org/gonum.org/v1/plot/vg) has support for a few backends in addition to PNG: JPEG, TIFF, EPS, PDF, SVG, TeX.
+
+The complete code for `mc-pi-plot.go` is here:
```

Comment:
maybe it's more readable as a link to your repo?

Reply:
Good point.
I'll link to a gist.

Reply:
done.